src/data-model-providers/codegen/CodegenDataModelProvider.cpp

```diff
@@ -553,165 +420,143 @@ std::optional<unsigned> CodegenDataModelProvider::TryFindEndpointIndex(EndpointI
-
-    VerifyOrReturnValue(cluster != nullptr, std::nullopt);
-
-    auto info = LoadClusterInfo(path, *cluster);
-
-    if (CHIP_ERROR * err = std::get_if<CHIP_ERROR>(&info))
+    for (auto & attribute : attributeSpan)
     {
+        err = result.Append(AttributeEntryFrom(path, attribute));
+        if (err != CHIP_NO_ERROR)
+        {
 #if CHIP_ERROR_LOGGING && CHIP_CONFIG_DATA_MODEL_EXTRA_LOGGING
-        ChipLogError(AppServer, "Failed to load cluster info: %" CHIP_ERROR_FORMAT, err->Format());
-#else
-        (void) err->Format();
+            ChipLogError(AppServer, "Failed to append attribute id: %" CHIP_ERROR_FORMAT, err.Format());
 #endif
-        return std::nullopt;
+            break;
+        }
     }
 
-    return std::make_optional(std::get<DataModel::ClusterInfo>(info));
+    return result;
 }
 
-ConcreteClusterPath CodegenDataModelProvider::FirstClientCluster(EndpointId endpointId)
+DataModel::MetadataList<ClusterId> CodegenDataModelProvider::ClientClusters(EndpointId endpointId)
 {
     const EmberAfEndpointType * endpoint = emberAfFindEndpointType(endpointId);
-    VerifyOrReturnValue(endpoint != nullptr, ConcreteClusterPath(endpointId, kInvalidClusterId));
-    VerifyOrReturnValue(endpoint->clusterCount > 0, ConcreteClusterPath(endpointId, kInvalidClusterId));
-    VerifyOrReturnValue(endpoint->cluster != nullptr, ConcreteClusterPath(endpointId, kInvalidClusterId));
 
-    return ConcreteClusterPath(endpointId, FirstClientClusterId(endpoint, 0, mClientClusterIterationHint));
-}
+    DataModel::MetadataList<ClusterId> result;
 
-ConcreteClusterPath CodegenDataModelProvider::NextClientCluster(const ConcreteClusterPath & before)
-{
-    // TODO: This search still seems slow (ember will loop). Should use index hints as long
-    //       as ember API supports it
-    const EmberAfEndpointType * endpoint = emberAfFindEndpointType(before.mEndpointId);
+    VerifyOrReturnValue(endpoint != nullptr, result);
+    VerifyOrReturnValue(endpoint->clusterCount > 0, result);
+    VerifyOrReturnValue(endpoint->cluster != nullptr, result);
+
+    const EmberAfCluster * begin = endpoint->cluster;
+    const EmberAfCluster * end   = endpoint->cluster + endpoint->clusterCount;
 
-    VerifyOrReturnValue(endpoint != nullptr, ConcreteClusterPath(before.mEndpointId, kInvalidClusterId));
-    VerifyOrReturnValue(endpoint->clusterCount > 0, ConcreteClusterPath(before.mEndpointId, kInvalidClusterId));
-    VerifyOrReturnValue(endpoint->cluster != nullptr, ConcreteClusterPath(before.mEndpointId, kInvalidClusterId));
+    size_t clientClusterCount = 0;
 
-    std::optional<unsigned> cluster_idx = TryFindClusterIndex(endpoint, before.mClusterId, ClusterSide::kClient);
-    if (!cluster_idx.has_value())
+    for (const EmberAfCluster * cluster = begin; cluster != end; cluster++)
     {
-        return ConcreteClusterPath(before.mEndpointId, kInvalidClusterId);
+        if (cluster->IsClient())
+        {
+            clientClusterCount++;
```

Comment:
`emberAfClusterCountForEndpointType(endpoint, /* server = */ false)`

Reply:
Updated ... but this was not public API before even though it is trivial ... so I had to add it to mocks and dynamic dispatcher...this feels off (too much duplication) and doing it right would mean figuring out a different library than attribute storage. I am concerned I am going down a rabbit hole if I start to update more.